pkg/controllers/machine/disruption/expiration.go

```diff
@@ -0,0 +1,83 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package disruption
+
+import (
+	"context"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/utils/clock"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
+	"github.com/aws/karpenter-core/pkg/utils/functional"
+	machineutil "github.com/aws/karpenter-core/pkg/utils/machine"
+)
+
+// Expiration is a machine sub-controller that adds or removes status conditions on expired machines based on TTLSecondsUntilExpired
+type Expiration struct {
+	kubeClient client.Client
+	clock      clock.Clock
+}
+
+//nolint:gocyclo
+func (e *Expiration) Reconcile(ctx context.Context, provisioner *v1alpha5.Provisioner, machine *v1alpha5.Machine) (reconcile.Result, error) {
+	hasExpiredCondition := machine.StatusConditions().GetCondition(v1alpha5.MachineExpired).IsTrue()
+
+	// From here there are three scenarios to handle:
+	// 1. If TTLSecondsUntilExpired is not configured, remove the expired status condition
+	if provisioner.Spec.TTLSecondsUntilExpired == nil {
+		if hasExpiredCondition {
+			_ = machine.StatusConditions().ClearCondition(v1alpha5.MachineExpired)
+			logging.FromContext(ctx).Debugf("removing expiration status condition from machine as expiration has been disabled")
+		}
+		return reconcile.Result{}, nil
+	}
+	node, err := machineutil.NodeForMachine(ctx, e.kubeClient, machine)
+	if machineutil.IgnoreNodeNotFoundError(err) != nil {
+		return reconcile.Result{}, err
```

Comment:
In the emptiness disruption controller, we remove the status condition if the node is not found. Why are we doing it differently here?

Reply:
We aren't doing it here since we don't technically need the node here to do this evaluation. We are basing expiration purely on the Machine; however, there is a middle timeframe where the Node may be around longer than the machine (due to migration) so we should handle this case by taking the older of the two of them. If we fail to get the Machine due to a NotFound error, we should just keep chugging along.

Reply:
Got it. Is there a point in the future where we'd remove this piece and the piece below, because we can enforce this invariant?

Reply:
Yes, we should remove this at the point that we introduce v1 and we rip out all the other stuff, including the machine migration mechanism. Let me make a comment in here with a `//TODO` to note that we should remove this in the future.